### DISCUSSION TITLE
Mehrere Viewerlayouts in einem Portal

### DISCUSSION THREAD

**Discussion by thomaskofler** [e-netze/webgis-community#14] [General]
Hallo zusammen :)

wir würden gerne für ein Projekt (WIS4Kids) mehrere viewerlayouts in einem Portal zur Verfügung haben, damit wir z.B. Logos in bestimmte Karten einfügen können. Momentan kann man nur ein viewerlayout-xml pro Portal verwenden.

LG Thomas

**Comment by gis-eni** [comment]
Unterschiedliche Viewerlayouts innerhalb einer Portalseite sind zur Zeit nicht vorgesehen. Allerdings ist in Zukunft geplant, dass der Anwender gerade im Desktopbereich unter verschieden Layouts wählen kann (zB Optimiert für Mausbedienung, Optimiert für Touchbedienung). Da wäre es dann theoretisch auch möglich ein Layout schon im Aufruf oder über die ``custom.js`` für eine Karte festzulegen. Diese Erweiterungen werden allerdings frühestens Ende Q1/2023 an Bord sein.

Jetzt gibt es für die Lösung des Problems folgende Möglichkeiten:

**Erstellung einer eigenen Portalseite**

Die entsprechende Karte wird auf einer neuen Portalseite gehostet. Damit der Aufruf von der gewohnten Portalseite (wie alle anderen Karten) aufrufbar ist, wir dort eine ``Redirect App`` mit dem *AppBuilder* erstellt. Die neue Karte erscheint dann auf der *Default* Portalseite als ganz normale Kachel, wie alle anderen Karten. Der Anwender wird beim Klick dann auf die Karte an das alternative Portal umgeleitet, wo das entsprechende Logo im Layout angefügt wurde.

Die Erstellung von Apps mit dem AppBuilder wird unter https://docs.webgiscloud.com/cloud/webgis_apps/appbuilder/index.html beschrieben. Das erste Beispiel ist das ``App Template: Redirect App``. Dort kann man einfach den Link zur Karten auf der alternativen Portalseite verlinken.

**Manipulation der Seite nach dem Start über ``custom.js``** 

Wird eine bestimmte Karte aufgerufen, kann das in der ``custom.js`` dazu genutzt werden, um das Layout anzupassen:

``` javascript
if (mapUrlName === "Basiskarte und Kataster") {
    webgis.events.on('viewer-map-created', function(channel, sender, map) {
        let $mylogo=$("<div>")
             .css({
                         position: "absolute",
                         left: 0, bottom: 0, width: 100, height: 100,
                         zIndex: 99999,
                         backgroundImage: "url(http://......../my-logo.png"
                   })
           .appendTo("#map")
    });
}
```

Hier wird, nachdem der Viewer geladen wurde (viewer-map-created) für die Karte ``Basiskarte und Kataster`` ein ``div``-Tag erzeugt und dem Kartenelement (``#map``) hinzugefügt. Über die Styles (``css``) kann Größe, Position und Bild angeben werden.

**Comment by thomaskofler** [comment]
Hallo Jürgen,

danke fürs schnelle Bearbeiten! Hab deinen Code in die custom.js übernommen, aber das Logo wird in der Karte nicht angezeigt,

hier mein Code (Karten im Portal = WIS4Kids):

``` javascript
if (mapUrlName === "WIS4Kids") {
    webgis.events.on('viewer-map-created', function(channel, sender, map) {
        let $mylogo=$("<div>")
             .css({
                         position: "absolute",
                         left: 0, bottom: 0, width: 100, height: 100,
                         zIndex: 99999,
                         backgroundImage: "url(https://www.salzburg.gv.at/sagisdaten_download/SAGIS_Daten/WIS4Kids/WIS_ein_Bein_white.png)"
                   })
           .appendTo("#map")
    });
}
```

LG Thomas

**Comment by gis-eni** [comment]
Welche WebGIS Version verwendest du? Das Event "viewer-map-created" wurde im letzten Release eingeführt, als euch gezeigt hab, wie man Vector Tile Caches in die Karte einfügen kann. Wenn du eine älter Version hast, wird der Code dann leider nie aufgerufen ;)

Du kannst probieren, ob der folgende Code funktioniert. Hier wird nicht das Event verwendet, sondern einfach einmal 5 Sekunden (=5000 Millisekunden) gewartet und davon ausgegangen, dass das Viewerlayout bis dahin bereit ist. Dann wird das Bild eingefügt:

```Javascript
if (mapUrlName === "WIS4Kids") {
    webgis.delayed(function() {
        let $mylogo=$("<div>")
             .css({
                         position: "absolute",
                         left: 0, bottom: 0, width: 100, height: 100,
                         zIndex: 99999,
                         backgroundImage: "url(https://www.salzburg.gv.at/sagisdaten_download/SAGIS_Daten/WIS4Kids/WIS_ein_Bein_white.png)"
                   })
           .appendTo("#map")
    },5000);
}
```

**Comment by thomaskofler** [comment]
Danke, funktioniert :)

LG Thomas
